doc/reference/kernel/timing/clocks.rst

```diff
@@ -1,161 +1,242 @@
+the specified value.  There are likewise microsecond, cycles and ticks
+variants of this API.
+
+Timing Internals
+================
+
+Timeout Queue
+-------------
+
+All Zephyr ``k_timeout_t`` events specified using the API above are
+managed in a single, global queue of events.  Each event is stored in
+a double-linked list, with an attendant delta count in ticks from the
+previous event.  The action to take on an event is specified as a
+callback function pointer provided by the subsystem requesting the
+event, along with a ``struct _timeout`` tracking struct that is
+expected to be embedded within subsystem-defined data structures (for
+example: a ``struct wait_q``, or a ``k_tid_t`` thread struct).
+
+Note that all variant units passed via a ``k_timeout_t`` are converted
+to ticks once on insertion into the list.  There no
+multiple-conversion steps internal to the kernel, so precision is
+guaranteed at the tick level no matter how many events exist or how
+long a timeout might be.
+
+Note that the list structure means that the CPU work involved in
+managing large numbers of timeouts is quadratic in the number of
+active timeouts.  The API design of the timeout queue was intended to
+permit a more scalable backend data structure, but no such
+implementation exists currently.
+
+Timer Drivers
+-------------
+
+Kernel timing at the tick level is driven by a timer driver with a
+comparatively simple API.
+
+* The driver is expected to be able to "announce" new ticks to the
+  kernel via the ``z_clock_nanounce()`` call, which passes an integer
+  number of ticks that have elapsed since the last announce call (or
+  system boot).  These calls can occur at any time, but the driver is
+  expected to attempt to ensure (to the extent practical given
+  interrupt latency interactions) that they occur near tick boundaries
+  (i.e. not "halfway through" a tick), and most importantly that they
+  be correct over time and subject to minimal skew vs. other counters
+  and real world time.
+
+* The driver is expected to provide a ``z_clock_set_timeout()`` call
+  to the kernel which indicates how many ticks may elapse before the
+  kernel must receive an announce call to trigger registered timeouts.
+  It is legal to announce new ticks before that moment (though they
+  must be correct) but delay after that will cause events to be
+  missed.  Note that the timeout value passed here is in a delta from
+  current time, but that does not absolve the driver of the
+  requirement to provide ticks at a steady rate over time.  Naive
+  implementations of this function are subject to bugs where the
+  fractional tick gets "reset" incorrectly and causes clock skew.
+
+* The driver is expected to provide a ``z_clock_elapsed()`` call which
+  provides a current indication of how many ticks have elapsed (as
+  compared to a real world clock) since the last call to
+  ``z_clock_announce()``, which the kernel needs to test newly
```

Comment:
Since last announced tick. The call to z_clock_announce() is always delayed (by interrupt/driver latency).